lib/ex_unit/lib/ex_unit/server.ex

```diff
@@ -107,26 +132,40 @@ defmodule ExUnit.Server do
     {:reply, diff, take_modules(%{state | loaded: :done})}
   end
 
-  def handle_call({:add, true, names}, _from, %{loaded: loaded} = state)
+  def handle_call({:add, {false = _async, _group}, names}, _from, %{loaded: loaded} = state)
+      when is_integer(loaded) do
+    state =
+      update_in(state.sync_modules, &Enum.reduce(names, &1, fn name, q -> :queue.in(name, q) end))
+
+    {:reply, :ok, state}
+  end
+
+  def handle_call({:add, {true = _async, nil = _group}, names}, _from, %{loaded: loaded} = state)
       when is_integer(loaded) do
     state =
       update_in(
         state.async_modules,
-        &Enum.reduce(names, &1, fn name, q -> :queue.in(name, q) end)
+        &Enum.reduce(names, &1, fn name, q -> :queue.in({:module, name}, q) end)
       )
 
-    {:reply, :ok, take_modules(state)}
+    {:reply, :ok, state}
   end
 
-  def handle_call({:add, false, names}, _from, %{loaded: loaded} = state)
+  def handle_call({:add, {true = _async, group}, names}, _from, %{loaded: loaded} = state)
       when is_integer(loaded) do
     state =
-      update_in(state.sync_modules, &Enum.reduce(names, &1, fn name, q -> :queue.in(name, q) end))
+      if Map.has_key?(state.async_groups, group) do
+        update_in(state.async_groups[group], &(names ++ &1))
+      else
+        state
+        |> put_in([:async_groups, group], names)
+        |> update_in([:async_modules], &:queue.in({:group, group}, &1))
+      end
```

Comment:
I would probably write this as:

```suggestion
    case state.async_groups do
      %{^group => entries} ->
        {%{async_groups | group => names ++ entries}, state.async_modules}
      %{} ->
        {Map.put(async_groups, group, names),
         :queue.in({:group, group}, state.async_modules)}
    end
```

Fewer traversals. :)